diff --git a/grammar.js b/grammar.js
index 4a0ff4b..4c0e897 100644
--- a/grammar.js
+++ b/grammar.js
@@ -111,7 +111,6 @@ module.exports = grammar({
     [$.array_expression],
     [$.visibility_modifier],
     [$.visibility_modifier, $.scoped_identifier, $.scoped_type_identifier],
-    [$.foreign_mod_item, $.function_modifiers],
   ],

   word: $ => $.identifier,
@@ -281,7 +280,8 @@ module.exports = grammar({
     ),

     foreign_mod_item: $ => seq(
-      optional('unsafe'),
+      optional($.visibility_modifier),
+      optional($._item_modifiers),
       $.extern_modifier,
       choice(
         ';',
@@ -406,6 +406,7 @@ module.exports = grammar({

     static_item: $ => seq(
       optional($.visibility_modifier),
+      optional(choice('unsafe', 'safe')),
       'static',

       // Not actual rust syntax, but made popular by the lazy_static crate.
@@ -458,12 +459,21 @@ module.exports = grammar({
       ';',
     ),

-    function_modifiers: $ => repeat1(choice(
+    function_modifiers: $ => choice(
+      $._item_modifiers,
+      $.extern_modifier,
+      seq(
+        $._item_modifiers,
+        $.extern_modifier,
+      ),
+    ),
+
+    _item_modifiers: _ => repeat1(choice(
       'async',
       'default',
       'const',
       'unsafe',
-      $.extern_modifier,
+      'safe',
     )),

     where_clause: $ => prec.right(seq(
@@ -1434,7 +1444,10 @@ module.exports = grammar({
         $.scoped_type_identifier,
       )),
       '{',
-      sepBy(',', choice($.field_pattern, $.remaining_field_pattern)),
+      sepBy(',', seq(
+        repeat($.attribute_item),
+        choice($.field_pattern, $.remaining_field_pattern),
+      )),
       optional(','),
       '}',
     ),
